tox.ini

```diff
@@ -0,0 +1,65 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+[flake8]
+max-line-length = 100
+
+[tox]
+skipsdist = True
+skip_missing_interpreters = True
+envlist = fmt, lint
+
+[vars]
+all_path = {[vars]src_path} {[vars]tst_path}
+src_path = {toxinidir}/
+tst_path = {toxinidir}/
+
+[testenv]
+passenv =
+    PYTHONPATH
+    CHARM_BUILD_DIR
+    MODEL_SETTINGS
+    KUBECONFIG
+setenv =
+    PYTHONPATH = {toxinidir}:{toxinidir}/lib:{[vars]src_path}
+    PYTHONBREAKPOINT=ipdb.set_trace
+    PY_COLORS=1
+
+[testenv:update-requirements]
+allowlist_externals =
+    bash
+    find
+    pip-compile
+    xargs
+commands =
+; uses 'bash -c' because piping didn't work in regular tox commands
+  pip-compile requirements.in
+  pip-compile requirements-fmt.in
+  bash -c 'find . -type f -name "requirements*.in" | xargs --replace=\{\} pip-compile --resolver=backtracking \{\}'
+deps =
+    pip-tools
+description = Update requirements files by executing pip-compile on all requirements*.in files, including those in subdirs.
+
+[testenv:fmt]
+commands =
+    isort {[vars]all_path}
+    black {[vars]all_path}
+deps =
+    -r requirements-fmt.txt
+description = Apply coding style standards to code
+
+[testenv:lint]
+commands =
+    # uncomment the following line if this charm owns a lib
+    # codespell {[vars]lib_path}
+    codespell {toxinidir}/. --skip {toxinidir}/./.git --skip {toxinidir}/./.tox \
+      --skip {toxinidir}/./build --skip {toxinidir}/./lib --skip {toxinidir}/./venv \
+      --skip {toxinidir}/./.mypy_cache \
+      --skip {toxinidir}/./icon.svg --skip *.json.tmpl
+    # pflake8 wrapper supports config from pyproject.toml
+    pflake8 {[vars]src_path}
+    isort --check-only --diff {[vars]all_path}
+    black --check --diff {[vars]all_path}
```

Comment:
I've moved recently to using `ruff` + `black` which I've  found much more stable in CI - it's also now the recommended path in the charmcraft template. See [this repo](https://github.com/jnsgruk/parca-operator) for example tox/pyproject file with equivalent rules

Reply:
@jnsgruk This linting process is the same CI template that is used across all KF charms. There is a conversation in-team and with COS to have a new unified CI (spec is in Braindump stage). We can address this in next sprint along with other dev process improvements.